File: MuseoPOO/restauracion.py
"""Módulo que define la clase Restauracion."""

from __future__ import annotations

from datetime import date
from typing import TYPE_CHECKING, Optional

if TYPE_CHECKING:
    from obra_de_arte import ObraDeArte


class Restauracion:
    """Representa un proceso de restauración de una obra."""

    def __init__(
        self,
        id_restauracion: int,
        obra: ObraDeArte,
        tipo_restauracion: str,
        fecha_inicio: date,
    ) -> None:
        self._id_restauracion = id_restauracion
        self._obra = obra
        self._tipo_restauracion = tipo_restauracion
        self._fecha_inicio = fecha_inicio
        self._fecha_fin: Optional[date] = None
        self._estado = "en_curso"

    def get_id_restauracion(self) -> int:
        return self._id_restauracion

    def get_obra(self) -> ObraDeArte:
        return self._obra

    def get_tipo_restauracion(self) -> str:
        return self._tipo_restauracion

    def get_fecha_inicio(self) -> date:
        return self._fecha_inicio

    def get_fecha_fin(self) -> Optional[date]:
        return self._fecha_fin

    def get_estado(self) -> str:
        return self._estado

    def registrar_inicio(self) -> None:
        """Registra el inicio de la restauración."""
        self._obra.set_estado("en_restauracion")
        self._estado = "en_curso"
        print(
            f"Restauración iniciada para "
            f"'{self._obra.get_titulo()}' "
            f"- Tipo: {self._tipo_restauracion}."
        )

    def registrar_fin(self, fecha_fin: Optional[date] = None) -> None:
        """Registra la finalización de la restauración."""
        if self._estado != "en_curso":
            print("Esta restauración ya fue finalizada.")
            return
        self._fecha_fin = fecha_fin or date.today()
        self._estado = "finalizada"
        self._obra.set_estado("expuesta")
        print(
            f"Restauración finalizada para "
            f"'{self._obra.get_titulo()}' el {self._fecha_fin}."
        )

    def __str__(self) -> str:
        return (
            f"Restauracion(id={self._id_restauracion}, "
            f"obra='{self._obra.get_titulo()}', "
            f"tipo='{self._tipo_restauracion}', "
            f"estado='{self._estado}')"
        )

File: Museo Java/Restauracion.java
import java.time.LocalDate;

public class Restauracion {

    private int idRestauracion;
    private ObraDeArte obra;
    private String tipoRestauracion;
    private LocalDate fechaInicio;
    private LocalDate fechaFin;
    private String estado;

    public Restauracion(int idRestauracion, ObraDeArte obra,
                        String tipoRestauracion, LocalDate fechaInicio) {
        this.idRestauracion = idRestauracion;
        this.obra = obra;
        this.tipoRestauracion = tipoRestauracion;
        this.fechaInicio = fechaInicio;
        this.fechaFin = null;
        this.estado = "en_curso";
    }

    public int getIdRestauracion() { return idRestauracion; }
    public ObraDeArte getObra() { return obra; }
    public String getTipoRestauracion() { return tipoRestauracion; }
    public LocalDate getFechaInicio() { return fechaInicio; }
    public LocalDate getFechaFin() { return fechaFin; }
    public String getEstado() { return estado; }

    public void registrarInicio() {
        obra.setEstado("en_restauracion");
        this.estado = "en_curso";
        System.out.println("Restauración iniciada para '" + obra.getTitulo() +
                "' - Tipo: " + tipoRestauracion + ".");
    }

    public void registrarFin(LocalDate fechaFin) {
        if (!"en_curso".equals(estado)) {
            System.out.println("Esta restauración ya fue finalizada.");
            return;
        }
        this.fechaFin = (fechaFin != null) ? fechaFin : LocalDate.now();
        this.estado = "finalizada";
        obra.setEstado("expuesta");
        System.out.println("Restauración finalizada para '" + obra.getTitulo() +
                "' el " + this.fechaFin + ".");
    }

    public void registrarFin() {
        registrarFin(null);
    }

    @Override
    public String toString() {
        return "Restauracion(id=" + idRestauracion + ", obra='" + obra.getTitulo() +
               "', tipo='" + tipoRestauracion + "', estado='" + estado + "')";
    }
}
